client/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -0,0 +1,468 @@
+     */
+    private Iterable<HttpHost> nextHost() {
+        Set<HttpHost> filteredHosts = new HashSet<>(hosts);
+        for (Map.Entry<HttpHost, DeadHostState> entry : blacklist.entrySet()) {
+            if (System.nanoTime() - entry.getValue().getDeadUntilNanos() < 0) {
+                filteredHosts.remove(entry.getKey());
+            }
+        }
+
+        if (filteredHosts.isEmpty()) {
+            //last resort: if there are no good hosts to use, return a single dead one, the one that's closest to being retried
+            List<Map.Entry<HttpHost, DeadHostState>> sortedHosts = new ArrayList<>(blacklist.entrySet());
+            Collections.sort(sortedHosts, new Comparator<Map.Entry<HttpHost, DeadHostState>>() {
+                @Override
+                public int compare(Map.Entry<HttpHost, DeadHostState> o1, Map.Entry<HttpHost, DeadHostState> o2) {
+                    return Long.compare(o1.getValue().getDeadUntilNanos(), o2.getValue().getDeadUntilNanos());
+                }
+            });
+            HttpHost deadHost = sortedHosts.get(0).getKey();
+            logger.trace("resurrecting host [" + deadHost + "]");
+            return Collections.singleton(deadHost);
+        }
+
+        List<HttpHost> rotatedHosts = new ArrayList<>(filteredHosts);
+        //TODO is it possible to make this O(1)? (rotate is O(n))
+        Collections.rotate(rotatedHosts, rotatedHosts.size() - lastHostIndex.getAndIncrement());
+        return rotatedHosts;
+    }
+
+    /**
+     * Called after each successful request call.
+     * Receives as an argument the host that was used for the successful request.
+     */
+    private void onSuccess(HttpHost host) {
+        DeadHostState removedHost = this.blacklist.remove(host);
+        if (logger.isDebugEnabled() && removedHost != null) {
+            logger.debug("removed host [" + host + "] from blacklist");
+        }
+    }
+
+    /**
+     * Called after each failed attempt.
+     * Receives as an argument the host that was used for the failed attempt.
+     */
+    private void onFailure(HttpHost host) throws IOException {
+        while(true) {
+            DeadHostState previousDeadHostState = blacklist.putIfAbsent(host, DeadHostState.INITIAL_DEAD_STATE);
+            if (previousDeadHostState == null) {
+                logger.debug("added host [" + host + "] to blacklist");
+                break;
+            }
+            if (blacklist.replace(host, previousDeadHostState, new DeadHostState(previousDeadHostState))) {
+                logger.debug("updated host [" + host + "] already in blacklist");
+                break;
+            }
+        }
+        failureListener.onFailure(host);
+    }
+
+    /**
+     * Sets a {@link FailureListener} to be notified each and every time a host fails
+     */
+    public synchronized void setFailureListener(FailureListener failureListener) {
```

Comment:
I replied to the same comment [here](https://github.com/elastic/elasticsearch/pull/18735#discussion_r66456178). I am not sure how we this would work given that Sniffer depends on RestClient and not the other way around. The Sniffer, when registered, has to set itself as a failure listener. How can it do that if failure listener is required when creating the RestClient?